### DISCUSSION TITLE
Post-Quantum Key Agreement in TLS 1.3

### DISCUSSION THREAD

**Discussion by reneme** [randombit/botan#3747] [Show and tell]
With [Botan 3.2.0 released yesterday](https://github.com/randombit/botan/blob/3.2.0/news.rst) we shipped experimental support for Post-Quantum Key Agreement in TLS 1.3 using Kyber as defined in [draft-ietf-tls-hybrid-design](https://datatracker.ietf.org/doc/html/draft-ietf-tls-hybrid-design).

Applications using the library's TLS implementation may enable PQC support simply by extending their TLS policy file (`key_exchange_groups`) with one or more of the supported key exchange schemes. Or [set up their override of `Botan::TLS::Callbacks::key_exchange_groups()](https://botan.randombit.net/handbook/api_ref/tls.html#_CPPv4NK3TLS6Policy19key_exchange_groupsEv) accordingly. The identifiers are compatible with the ones defined by [OQS](https://github.com/open-quantum-safe/oqs-provider/blob/main/ALGORITHMS.md) and [Cloudflare](https://pq.cloudflareresearch.com/). Support for FrodoKEM [will be added](https://github.com/randombit/botan/pull/3679) in a Botan 3.3.0.

| Policy algorithm name            | `Botan::TLS::Group_Params::...`         | TLS identifier |
| -------------------------------- | --------------------------------------- | -------------- |
| `Kyber-512-r3`                   | `KYBER_512_R3_OQS`                      | `0x023A`       |
| `Kyber-768-r3`                   | `KYBER_768_R3_OQS`                      | `0x023C`       |
| `Kyber-1024-r3`                  | `KYBER_1024_R3_OQS`                     | `0x023D`       |
| `x25519/Kyber-512-r3/cloudflare` | `HYBRID_X25519_KYBER_512_R3_CLOUDFLARE` | `0xFE30`       |
| `x25519/Kyber-512-r3`            | `HYBRID_X25519_KYBER_512_R3_OQS`        | `0x2F39`       |
| `x25519/Kyber-768-r3`            | `HYBRID_X25519_KYBER_768_R3_OQS`        | `0x6399`       |
| `secp256r1/Kyber-512-r3`         | `HYBRID_SECP256R1_KYBER_512_R3_OQS`     | `0x2F3A`       |
| `secp256r1/Kyber-768-r3`         | `HYBRID_SECP256R1_KYBER_768_R3_OQS`     | `0x639A`       |
| `secp384r1/Kyber-768-r3`         | `HYBRID_SECP384R1_KYBER_768_R3_OQS`     | `0x2F3C`       |
| `secp521r1/Kyber-1024-r3`        | `HYBRID_SECP521R1_KYBER_1024_R3_OQS`    | `0x2F3D`       |

For a simple test, just [compile Botan's CLI tool](https://botan.randombit.net/handbook/building.html), create a TLS policy file (`pqc_policy.txt`):

```
allow_tls13 = true
allow_tls12 = false
key_exchange_groups = x25519/Kyber-768-r3
```

and run the CLI like this:

```bash
./botan tls_client --debug --policy=./pqc_policy.txt cloudflare.com
```

to perform a TLS 1.3 handshake secured by a hybrid of Kyber and X25519.

TLS with post-quantum key exchange is still experimental with both the protocol and the asymmetric algorithms not being published as final standards, yet. To ensure compatibility, we added [an automated test](https://github.com/randombit/botan/pull/3732) (executed nightly) with the experimental deployments at Cloudflare, IBM, Google and OpenQuantumSafe.

**Comment by RB-UK-BMS** [comment]
Hello - i would like to use a different Quantum Key than Kyber from another origin.
This is for a commercial IOT application.
Would someone please DM me richard@Bluemeshsolutions.com
Many thanks - RB
